src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java

```diff
@@ -501,4 +519,303 @@ public int hashCode() {
+
+		//TODO: a simple envelope test
+		
+		OGCConcreteGeometryCollection flattened1 = flatten();
+		if (flattened1.isEmpty())
+			return true;
+		OGCConcreteGeometryCollection otherCol = new OGCConcreteGeometryCollection(another, esriSR);
+		OGCConcreteGeometryCollection flattened2 = otherCol.flatten();
+		if (flattened2.isEmpty())
+			return true;
+		
+		for (int i = 0, n2 = flattened2.numGeometries(); i < n2; ++i) {
+			OGCGeometry g2 = flattened2.geometryN(i);
+			boolean good = false;
+			for (int j = 0, n1 = flattened1.numGeometries(); j < n1; ++j) {
+				OGCGeometry g1 = flattened1.geometryN(i);
+				if (g1.contains(g2)) {
+					good = true;
+					break;
+				}
+			}
+			
+			if (!good)
+				return false;
+		}
+		
+		//each geometry of another is contained in a geometry from this.
+		return true;
+	}
+	
+	/**
+	 * Checks if collection is flattened.
+	 * @return True for the flattened collection. A flattened collection contains up to three non-empty geometries:
+	 * an OGCMultiPoint, an OGCMultiPolygon, and an OGCMultiLineString.
+	 */
+	public boolean isFlattened() {
+		int n = numGeometries();
+		if (n > 3)
+			return false;
+
+		for (int i = 0; i < n; ++i) {
+			OGCGeometry g = geometryN(i);
+			if (g.isEmpty())
+				return false;//no empty allowed
+			
+			String t = g.geometryType();
+			if (t != OGCMultiPoint.TYPE && t != OGCMultiPolygon.TYPE && t != OGCMultiLineString.TYPE)
+				return false;
+		}
+		
+		return true;
+	}
+
+	/**
+	 * Flattens Geometry Collection.
+	 * The result collection contains up to three geometries:
+	 * an OGCMultiPoint, an OGCMultiPolygon, and an OGCMultilineString.
+	 * @return A flattened Geometry Collection, or self if already flattened.
+	 */
+	public OGCConcreteGeometryCollection flatten() {
+		if (isFlattened()) {
+			return this;
+		}
+		
+		OGCMultiPoint multiPoint = null;
+		ArrayList<Geometry> polygons = null;
+		OGCMultiLineString polyline = null;
+		GeometryCursor gc = getEsriGeometryCursor();
+		for (Geometry g = gc.next(); g != null; g = gc.next()) {
+			if (g.isEmpty())
+				continue;
+			
+			Geometry.Type t = g.getType();
+
+			if (t == Geometry.Type.Point) {
+				if (multiPoint == null) {
+					multiPoint = new OGCMultiPoint(esriSR);
+				}
+
+				((MultiPoint)multiPoint.getEsriGeometry()).add((Point)g);
+				continue;
+			}
+
+			if (t == Geometry.Type.MultiPoint) {
+				if (multiPoint == null)
+					multiPoint = new OGCMultiPoint(esriSR);
+				
+				((MultiPoint)multiPoint.getEsriGeometry()).add((MultiPoint)g, 0, -1);
+				continue;
+			}
+			
+			if (t == Geometry.Type.Polyline) {
+				if (polyline == null)
+					polyline = new OGCMultiLineString(esriSR);
+				
+				((MultiPath)polyline.getEsriGeometry()).add((Polyline)g, false);
+				continue;
+			}
+			
+			if (t == Geometry.Type.Polygon) {
+				if (polygons == null)
+					polygons = new ArrayList<Geometry>();
+				
+				polygons.add(g);
+				continue;
+			}
+			
+			throw new GeometryException("internal error");//what else?
```

Comment:
perhaps, clarify that unexpected type was encountered

`throw new GeometryException("Internal error: unexpected geometry type: " + t);/`